lib/crx.js

```diff
@@ -0,0 +1,164 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import * as crxpb from './crx3.pb.js'
+import fs from 'fs-extra'
+import crypto from 'crypto'
+import JSZip from 'jszip'
+import glob from 'glob'
+import path from 'path'
+import util from './util.js'
+import zlib from 'zlib'
+import Pbf from 'pbf'
+
+const CRX_ID_SIZE = 16
+const CRX_FILE_MAGIC = 'Cr24'
+const CRX_VERSION = 3
+const CRX_SIGNED_DATA_HEADER = Buffer.from('CRX3 SignedData\x00', 'utf-8')
+
+const toLE32 = (num) => {
+  const buffer = Buffer.alloc(4)
+  buffer.writeUInt32LE(num)
+  return buffer
+}
+
+const createCrxId = (signingKey) => {
+  const hash = crypto.createHash('sha256')
+  const publicKey = crypto.createPublicKey(signingKey)
+  hash.update(
+    publicKey.export({
+      type: 'spki',
+      format: 'der'
+    })
+  )
+  return hash.digest().subarray(0, CRX_ID_SIZE)
+}
+
+const createSignedData = (id) => {
+  const pbf = new Pbf()
+  crxpb.writeSignedData({ crx_id: id }, pbf)
+  return pbf.finish()
+}
+
+const signContentAndCreateHeader = (content, extensionKey, publisherKeys) => {
+  const crxId = createCrxId(extensionKey)
+
+  const header = {
+    sha256_with_rsa: [],
+    signed_header_data: createSignedData(crxId)
+  }
+
+  for (const signingKey of [extensionKey, ...publisherKeys]) {
+    const signer = crypto.createSign('sha256')
+
+    // Assemble SignedData section.
+    signer.update(CRX_SIGNED_DATA_HEADER)
+    signer.update(toLE32(header.signed_header_data.length))
+    signer.update(header.signed_header_data)
+    signer.update(content)
+
+    const publicKey = crypto.createPublicKey(signingKey)
+    const proof = {
+      public_key: publicKey.export({ type: 'spki', format: 'der' }),
+      signature: signer.sign(signingKey)
+    }
+    // Add signature.
+    header.sha256_with_rsa.push(proof)
+  }
+  return header
+}
+
+const zipContent = async (stagingDir) => {
+  const processFiles = (dir, callback) => {
+    const files = glob.sync('**/*', {
+      nodir: true,
+      absolute: false,
+      cwd: dir
+    })
+
+    for (const relativePath of files) {
+      callback(relativePath)
+    }
+  }
+
+  const zip = new JSZip()
+  processFiles(stagingDir, (file) => {
+    zip.file(file, fs.readFileSync(path.join(stagingDir, file)))
+  })
+
+  return await zip.generateAsync({
+    type: 'Uint8Array',
+    compression: 'DEFLATE',
+    compressionOptions: { level: 9 }
+  })
+}
+
+const updateManifest = (stagingDir, extensionKey) => {
+  const manifestFile = path.join(stagingDir, 'manifest.json')
+  const manifest = util.parseManifest(manifestFile)
+  const publicKey = crypto.createPublicKey(extensionKey)
+
+  // Public key which used to generate the extension/crx id.
+  manifest.key = publicKey
+    .export({
+      type: 'spki',
+      format: 'der'
+    })
+    .toString('base64')
+
+  // The WebStore update URL makes the browser think that the extension is from the Web Store.
+  manifest.update_url = 'https://clients2.google.com/service/update2/crx'
```

Comment:
This is likely result in sending update requests to Chrome Store. 
How we going to update these extensions?

Reply:
all update requests go to our go-updater service